internal/builder/docker.go

```diff
@@ -0,0 +1,130 @@
+)
+
+const (
+	// DefaultDockerImage is the remote docker image
+	// name used by default
+	DefaultDockerImage = "lucor/fyne-cross:1.4.0"
+)
+
+type dockerImageProvider interface {
+	DockerImageName() (string, error)
+}
+
+type dockerRemoteImage struct {
+	Name string
+}
+
+func (d *dockerRemoteImage) DockerImageName() (string, error) {
+	return d.Name, nil
+}
+
+type dockerFile struct {
+	Path string
+}
+
+func (d *dockerFile) DockerImageName() (string, error) {
+	// Hash the Dockerfile to produce an image name
+	data, err := ioutil.ReadFile(d.Path)
+	if err != nil {
+		return "", err
+	}
+	h := sha1.Sum(data)
+	image := hex.EncodeToString(h[:])
+	dir := filepath.Dir(d.Path)
+	cmd := exec.Command("docker", "build", dir, "-t", image)
+	if err := cmd.Run(); err != nil {
+		return "", err
+	}
+	return image, nil
+}
+
+var (
+	dockerProvider dockerImageProvider = &dockerRemoteImage{Name: DefaultDockerImage}
+)
+
+func isFile(p string) bool {
+	st, err := os.Stat(p)
+	return err == nil && !st.IsDir()
+}
+
+func SetDockerImage(imageName string) {
+	if filepath.Base(imageName) == "Dockerfile" && isFile(imageName) {
+		dockerProvider = &dockerFile{Path: imageName}
+	} else {
+		dockerProvider = &dockerRemoteImage{Name: imageName}
+	}
+}
+
+// dockerCmd exec a command inside the container for the specified image
+func dockerCmd(image string, vol *volume.Volume, env []string, workDir string, command []string, verbose bool) *exec.Cmd {
+	// define workdir
+	w := vol.WorkDirContainer()
+	if workDir != "" {
+		w = workDir
+	}
+
+	args := []string{
+		"run", "--rm", "-t",
+		"-w", w, // set workdir
+		"-v", fmt.Sprintf("%s:%s", vol.WorkDirHost(), vol.WorkDirContainer()), // mount the working dir
+		"-v", fmt.Sprintf("%s:%s", vol.CacheDirHost(), vol.CacheDirContainer()), // mount the cache dir
+		"-e", "CGO_ENABLED=1", // enable CGO
+		"-e", fmt.Sprintf("GOCACHE=%s", vol.GoCacheDirContainer()), // mount GOCACHE to reuse cache between builds
+	}
+
+	// add custom env variables
+	for _, e := range env {
+		args = append(args, "-e", e)
+	}
+
+	// attempt to set fyne user id as current user id to handle mount permissions
+	// on linux and MacOS
+	if runtime.GOOS != "windows" {
+		u, err := user.Current()
+		if err == nil {
+			args = append(args, "-e", fmt.Sprintf("fyne_uid=%s", u.Uid))
+		}
+	}
+
+	// specify the image to use
+	args = append(args, image)
+
+	// add the command to execute
+	args = append(args, command...)
+
+	// run the command inside the container
+	cmd := exec.Command("docker", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if verbose {
+		fmt.Println(cmd.String())
+	}
+
+	return cmd
+}
+
+// runDockerCmd is a shorthand for running dockerCmd() with the current docker
+// image name provider and then calling Run on the returned command
+func runDockerCmd(vol *volume.Volume, env []string, workDir string, command []string, verbose bool) error {
+	// retrieve image name
+	image, err := dockerProvider.DockerImageName()
```

Comment:
If possible, it is preferable to have all dependencies declared as function argument to allow test / mocking.

Reply:
This function seems to take already too many arguments. Plus, it's also very small (`dockerCmd()` does most of the work) and you can still test by changing the package-level docker image beforehand. What I hadn't consider is that at some point you might want to run different builds in parallel, and in that case package-level state definitely gets on the way.

Reply:
Yes, run builds in parallel could be something we'd want to add in future.
Anyway, since we are going to add the `DockerImage` method into the builder interface what about to use only the dockerCmd passing into it the image as b.DockerImage() ?
It won't reduce the number of arguments but will self document the required dependencies.

Reply:
With the latest commits all package level state should now be gone. All state is now encapsulated inside each Builder implementation.

Reply:
Indeed, but what I meant is that `dockerProvider` was not an explict dependency since it not passed as function argument :-)